### PR TITLE
Add source zh.manhuaren

### PR DESCRIPTION
**Pull request by geniusgordon**
Checklist:
- [x] Updated source's version for individual source changes
- [x] Updated all sources' versions for template changes
- [x] Set appropriate `nsfw` value
- [x] Did not change `id` even if a source's name or language were changed
- [x] Tested the modifications by running it on the simulator or a test device

### CODE REVIEW

**Comment by hyspace** [comment]
Hi, what is being blocked？

**Comment by hyspace** [comment]
It seems like the icon is not transparent lol

File: src/rust/zh.manhuaren/src/helper.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+	temp.push_str("GET");
+
+	for a in args {
+		temp.push_str(&a.0);
+		temp.push_str(&encode_uri_component(String::from(&a.1)));
+	}
+
+	temp.push_str(GSN_KEY);
+
+	print(&temp);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/helper.rs:123]: best not to leave print statements in

File: src/rust/zh.manhuaren/src/helper.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+pub fn generate_get_query(args: &mut Vec<(String, String)>) -> String {
+	args.push((String::from("gak"), String::from("android_manhuaren2")));
+
+	let gsn = generate_gsn_hash(args);
+
+	args.push((String::from("gsn"), String::from(&gsn)));
+
+	print("gsn:");
+	print(gsn);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/helper.rs:136]: remove

File: src/rust/zh.manhuaren/src/helper.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+		let v = encode_uri_component(String::from(&a.1));
+		qs.push_str(&format(format_args!("{}={}", a.0, v)));
+	}
+
+	qs
+}
+
+pub fn request<T: AsRef<str>>(url: T, method: HttpMethod) -> Result<String> {
+	let req = Request::new(url, method).header("X-Yq-Yqci", "{\"le\": \"zh\"}");
+	req.string()
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/helper.rs:154]: could one-line this

File: src/rust/zh.manhuaren/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+			}
+			_ => continue,
+		}
+	}
+
+	print(format(format_args!("page: {}", page)));
+	print(format(format_args!("query: {}", query)));
+	print(format(format_args!("status: {}", status)));
+	print(format(format_args!("genre: {}", genre)));
+	print(format(format_args!("sort: {}", sort)));
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:66]: remove

File: src/rust/zh.manhuaren/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+		get_manga_list_by_query(query, page)
+	}
+}
+
+#[get_manga_listing]
+fn get_manga_listing(_: Listing, _: i32) -> Result<MangaPageResult> {
+	print("get_manga_listing");
+
+	todo!()
+}
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:87]: there are no listings, so this doesn't need to be implemented. just remove it.

File: src/rust/zh.manhuaren/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+#[get_manga_details]
+fn get_manga_details(id: String) -> Result<Manga> {
+	let mut args: Vec<(String, String)> = Vec::new();
+
+	args.push((String::from("mangaId"), id));
+
+	let qs = helper::generate_get_query(&mut args);
+
+	print("qs:");
+	print(&qs);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:98]: remove

File: src/rust/zh.manhuaren/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+
+	let url = String::from(API_URL) + "/v1/manga/getDetail?" + &qs;
+	print("url:");
+	print(&url);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:102]: remove

File: src/rust/zh.manhuaren/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+
+	let body = helper::request(url, HttpMethod::Get)?;
+	// print(&body);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:105]: remove

File: src/rust/zh.manhuaren/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+
+	let json = json::parse(body)?.as_object()?;
+	let manga = json.get("response").as_object()?;
+
+	let category_str = manga
+		.get("mangaTheme")
+		.as_string()
+		.unwrap_or_else(|_| StringRef::from(""))
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:113]: I think you can just do this?
```suggestion
		.unwrap_or(StringRef::from(""))
```

File: src/rust/zh.manhuaren/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+		.read();
+
+	let categories: Vec<&str> = category_str.split(' ').collect();
+
+	Ok(Manga {
+		id: helper::i32_to_string(manga.get("mangaId").as_int().unwrap_or(0) as i32),
+		cover: manga
+			.get("mangaPicimageUrl")
+			.as_string()
+			.unwrap_or_else(|_| StringRef::from("-"))
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:123]: these should use empty strings if they don't exist, not "-"

File: src/rust/zh.manhuaren/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+
+	print("qs:");
+	print(&qs);
+
+	let url = String::from(API_URL) + "/v1/manga/getDetail?" + &qs;
+	print("url:");
+	print(&url);
+
+	let body = helper::request(url, HttpMethod::Get)?;
+	// print(&body);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:177]: remove prints

File: src/rust/zh.manhuaren/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+
+	chapter_arr.append(&mut parse_chapters(&manga, "mangaEpisode"));
+	chapter_arr.append(&mut parse_chapters(&manga, "mangaWords"));
+	chapter_arr.append(&mut parse_chapters(&manga, "mangaRolls"));
+
+	Ok(chapter_arr)
+}
+
+#[get_page_list]
+fn get_page_list(manga_id: String, chapter_id: String) -> Result<Vec<Page>> {
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:192]: remove prints

File: src/rust/zh.manhuaren/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+	Ok(parse_page(&manga))
+}
+
+#[modify_image_request]
+fn modify_image_request(_: Request) {}
+
+#[handle_url]
+fn handle_url(_: String) -> Result<DeepLink> {
+	todo!()
+}
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:229]: these don't need to be implemented

File: src/rust/zh.manhuaren/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,512 @@
...
+
+fn get_manga_list_by_filter(filter: ListFilter, page: i32) -> Result<MangaPageResult> {
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:231]: remove prints

File: src/rust/zh.manhuaren/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+
+	for manga in mangas {
+		let manga_obj = manga.as_object()?;
+
+		manga_arr.push(Manga {
+			id: helper::i32_to_string(manga_obj.get("mangaId").as_int().unwrap_or(0) as i32),
+			cover: manga_obj
+				.get("mangaCoverimageUrl")
+				.as_string()
+				.unwrap_or_else(|_| StringRef::from("-"))
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:351]: again, use empty strings here instead

File: src/rust/zh.manhuaren/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+			let mut page_arr: Vec<Page> = Vec::new();
+
+			for (i, p) in pages.enumerate() {
+				let p_str = p.as_string().unwrap_or_else(|_| StringRef::from("")).read();
+
+				let mut url = helper::encode_uri(String::from(&host));
+				url.push_str(&p_str);
+				url.push_str(&query);
+
+				// print(&url);
```
Comment (Skittyblock) [inline src/rust/zh.manhuaren/src/lib.rs:493]: remove

**Review by Skittyblock — CHANGES_REQUESTED** [review]
also make sure to run `cargo +nightly clippy` for other changes you should make, and `cargo +nightly fmt` to correct some formatting.

**Comment by Skittyblock** [comment]
replaced by #212

**Pull request closed without merging**